Support multiple natively-routed CIDRs in ENI mode

twpayne opened the pull request.

tgraf commented on pkg/aws/eni/types/types.go:
```diff
@@ -200,6 +200,11 @@ type ENIStatus struct {
 	// +optional
 	ENIs map[string]ENI `json:"enis,omitempty"`
+
+	// Subnets is the list of subnets
```
```
Subnets is the list of subnets in the VPC and the availability zone to which the node is attached to
```

tgraf commented on pkg/node/types/node.go:
```diff
@@ -81,6 +87,49 @@ func ParseCiliumNode(n *ciliumv2.CiliumNode) (node Node) {
+		endpointAddresses := make([]Address, 0, len(eni.Addresses))
+		for _, address := range eni.Addresses {
+			endpointAddress := Address{
+				Type: addressing.NodeInternalIP, // FIXME is this correct?
```
We could probably just use `net.IP` instead of `Address` as we'll never use the `Type`.

tgraf commented on pkg/datapath/linux/node.go:
```diff
@@ -903,7 +905,12 @@ func (n *linuxNodeHandler) nodeUpdate(oldNode, newNode *nodeTypes.Node, firstAdd
+			// call to cidr.DiffCIDRLists in updateOrRemoveNodeRoutes.
+			oldIPv4CIDRs := append([]*cidr.CIDR{oldIP4Cidr}, oldIPv4NativeCIDRs...)
+			newIPv4CIDRs := append([]*cidr.CIDR{newNode.IPv4AllocCIDR}, newNode.IPv4NativeRoutingCIDRs...)
+			n.updateOrRemoveNodeRoutes(oldIPv4CIDRs, newIPv4CIDRs, isLocalNode)
```
This doesn't make sense, `updateOrRemoveNodeRoutes()` inserts routes pointing to the `cilium_host` interface. We'll need to update `ip rules` as done by:
```
func (info *RoutingInfo) Configure(ip net.IP, mtu int, compat bool) error {
```

tgraf commented on pkg/datapath/linux/node.go:
```diff
@@ -946,9 +951,246 @@ func (n *linuxNodeHandler) nodeUpdate(oldNode, newNode *nodeTypes.Node, firstAdd
+
+		// Configure new interfaces.
+		for _, addedInterface := range addedInterfaces {
+			mtu := 1500 // FIXME where do we get the MTU from?
```
```
n.nodeConfig.MtuConfig.GetDeviceMTU()
```

tgraf commented on pkg/datapath/linux/node.go:
```diff
@@ -946,9 +951,246 @@ func (n *linuxNodeHandler) nodeUpdate(oldNode, newNode *nodeTypes.Node, firstAdd
+		addedRules, removedRules := diffRules(oldRules, newRules)
+		for _, addedRule := range addedRules {
+			if err := route.ReplaceRule(*addedRule); err != nil {
+				return fmt.Errorf("unable to install ip rule: %s", err)
```
Does it make sense to error out here? We should probably attempt to add as many rules as possible instead and log errors. The node handler erroring out doesn't really block anything. This *must* succeed so we should probably rather retry instead.

tgraf commented on pkg/datapath/linux/node.go:
```diff
@@ -946,9 +951,246 @@ func (n *linuxNodeHandler) nodeUpdate(oldNode, newNode *nodeTypes.Node, firstAdd
+		for _, addedInterface := range addedInterfaces {
+			mtu := 1500 // FIXME where do we get the MTU from?
+			if _, err := linuxrouting.RetrieveIfIndexFromMAC(addedInterface.MAC, mtu); err != nil {
+				return err
```
We should not error out but just warn. We shouldn't skip the node handler because we already have the ifindex so we can continue.

tgraf reviewed (changes requested):
Looks very good. Added comment to derive MTU. The main change needed is error handling.

aanm commented on pkg/datapath/linux/node.go:
```diff
@@ -946,9 +951,246 @@ func (n *linuxNodeHandler) nodeUpdate(oldNode, newNode *nodeTypes.Node, firstAdd
+			tableID = linuxrouting.ComputeTableIDFromIfaceNumber(iface.Index)
+		}
+
+		sort.Sort(ipSlice(iface.EndpointAddresses))
```
why not use the `sort.Slice` here?

aanm reviewed (commented):
Only one small nit. Also, would it be possible to have more context in the commit messages?

tgraf commented on pkg/datapath/linux/node.go:
```diff
@@ -947,9 +952,285 @@ func (n *linuxNodeHandler) nodeUpdate(oldNode, newNode *nodeTypes.Node, firstAdd
+			routesToAdd, failedAddRoutes = failedAddRoutes, nil
+			routesToRemove, failedRemoveRoutes = failedRemoveRoutes, nil
+
+			// FIXME maybe add an exponential back-off?
```
I think it's fine given that we only retry 3 times.

tgraf commented on pkg/datapath/linux/node.go:
```diff
@@ -947,9 +952,285 @@ func (n *linuxNodeHandler) nodeUpdate(oldNode, newNode *nodeTypes.Node, firstAdd
+				IP:   net.IPv4zero,
+				Mask: net.CIDRMask(0, 32),
+			},
+			// Scope: netlink.SCOPE_LINK, // FIXME is this needed?
```
Nope

tgraf commented on pkg/datapath/linux/node.go:
```diff
@@ -947,9 +952,285 @@ func (n *linuxNodeHandler) nodeUpdate(oldNode, newNode *nodeTypes.Node, firstAdd
+
+		// Default route to the VPC or subnet gateway.
+		defaultRoute := &netlink.Route{
+			LinkIndex: iface.Index, // FIXME is this correct?
```
No, it will use the `LinkIndex` from the `/32` route just above

tgraf commented on cilium-health/launch/endpoint.go:
```diff
@@ -345,6 +345,7 @@ func LaunchAsEndpoint(baseCtx context.Context,
 		return nil, fmt.Errorf("Error while configuring routes: %s", err)
 	}
 
+	// FIXME is this still needed?
```
Nope, this can be removed

tgraf commented on pkg/datapath/linux/node.go:
```diff
@@ -947,9 +952,285 @@ func (n *linuxNodeHandler) nodeUpdate(oldNode, newNode *nodeTypes.Node, firstAdd
+// nodeIPNets returns all natively routed CIDRs for node.
+func nodeIPNets(node *nodeTypes.Node) []*net.IPNet {
+	var ipNets []*net.IPNet
+	if node.IPv4AllocCIDR != nil && node.IPv4AllocCIDR.IPNet != nil {
```
Nope, this can be removed

joestringer commented on cilium-health/launch/endpoint.go:
```diff
@@ -345,18 +344,6 @@ func LaunchAsEndpoint(baseCtx context.Context,
 		return nil, fmt.Errorf("Error while configuring routes: %s", err)
 	}
 
-	if option.Config.IPAM == ipamOption.IPAMENI {
```
If you're removing these bits, then the `routingConfig` now becomes unused and can be removed.

(Though I'll also naturally want to ask, if this code is not handled here then where is it handled? Maybe I'll figure that out later in the review..)

joestringer replied:
btw, here be dragons: We identified and manually tested the fix that introduced this code but I don't think there's any automated testing or regression checking.

twpayne replied:
Thanks for the heads-up. `routingConfig` removed, and yes, it seems that we have to rely on manual testing here.

joestringer commented on pkg/aws/eni/instances.go:
```diff
@@ -141,6 +141,21 @@ func (m *InstancesManager) FindSecurityGroupByTags(vpcID string, required ipamTy
 	return securityGroups
 }
 
+// GetMatchingSubnets returns all subnets in the given VPC and availability zone.
```
The commit message here refers to an ENU, which I can only assume is a hybrid of a large flightless bird and the network interfaces found deep in the jungles of the Amazon datacenters. Perhaps it was supposed to be ENI? ;-)

joestringer commented on pkg/datapath/linux/node.go:
```diff
@@ -947,9 +952,278 @@ func (n *linuxNodeHandler) nodeUpdate(oldNode, newNode *nodeTypes.Node, firstAdd
 }
 
+// nodeRulesAndRoutes returns the rules and routes required to configure node.
+// It is based on pkg/datapath/linux/routing.Configure.
```
When you say "based on", what is the overlap and how can we minimize code duplication?

twpayne replied:
`Configure` is still used for `ipam: azure`. The amount of overlap depends on the configuration. I would suggest that consolidating the code is best done in a follow-up PR.

christarazi replied:
+1, followup makes sense

joestringer reviewed (commented):
Do I follow right that the intent here is to move away from inline rules/routes setup for each pod during CNI ADD and more towards somehow orchestrating this setup via the operator which then notifies (each?) node to then react to the new IPs associated with the ENIs and configure the routing for all IPs on the node?

I'm struggling to see where the per-pod trigger for the update occurs now. How do we ensure that the networking is all set up and working for a new pod during CNI ADD before we return a successful operation (ie, "Yes, we have provisioned the network, here is your routes and IP" that we need to hand back as part of the CNI dance)?

christarazi commented on pkg/datapath/linux/node.go:
```diff
@@ -950,6 +959,273 @@ func (n *linuxNodeHandler) nodeUpdate(oldNode, newNode *nodeTypes.Node, firstAdd
+	for retry := 0; retry < maxRetries; retry++ {
+		for _, rule := range rulesToAdd {
+			if err := route.ReplaceRule(*rule); err != nil {
+				log.WithError(err).Errorf("add rule %s failed", rule)
```
I think we can be a bit more descriptive in these errors and use structured logging:

```
log.WithField(logfields.Rule, rule).WithError(err).Errorf("Failed to add routing rule in ENI IPAM mode")
```

This applies to the others below.

christarazi commented on pkg/datapath/linux/node.go:
```diff
@@ -950,6 +959,273 @@ func (n *linuxNodeHandler) nodeUpdate(oldNode, newNode *nodeTypes.Node, firstAdd
+	for _, addedInterface := range addedInterfaces {
+		mtu := n.nodeConfig.MtuConfig.GetDeviceMTU()
+		if _, err := linuxrouting.RetrieveIfIndexFromMAC(addedInterface.MAC, mtu); err != nil {
+			log.WithError(err).Errorf("Unable to configure interface index %d mac %s", addedInterface.Index, addedInterface.MAC)
```
We should use structured logging here:

```
log.WithFields({
  ...
}).WithError(...)
```

christarazi commented on plugins/cilium-cni/cilium-cni.go:
```diff
@@ -501,7 +501,7 @@ func cmdAdd(args *skel.CmdArgs) (err error) {
 
 	switch conf.IpamMode {
-	case ipamOption.IPAMENI, ipamOption.IPAMAzure:
+	case ipamOption.IPAMAzure:
```
I think it's worth giving more context in the commit msg how we've now migrated away from utilizing the CNI model for IPAM in ENI mode where routing was set up on a per-pod basis. Whereas with this PR, we move to a per-node routing setup.

christarazi commented on pkg/node/types/node.go:
```diff
@@ -44,7 +51,7 @@ func (nn Identity) String() string {
 
 // ParseCiliumNode parses a CiliumNode custom resource and returns a Node
-// instance. Invalid IP and CIDRs are silently ignored
+// instance. Invalid IPs, CIDRs, and MACs are silently ignored
```
It seems now that this code path is more crucial than it was before, does it make sense to revisit this and at least log about failures?

twpayne replied:
With the new approach, we determine the rules and routes directly from the `CiliumNode` CRD, so this change is no longer needed.

christarazi commented on pkg/node/types/node.go:
```diff
@@ -151,6 +204,16 @@ func (n *RegisterNode) DeepKeyCopy() store.LocalKey {
+	Gateway           Address
+	Index             int
+	MAC               mac.MAC
+	EndpointAddresses []net.IP
```
godocs on these fields

twpayne replied:
With the new approach, this struct is no longer needed.

christarazi commented on cilium-health/launch/endpoint.go:
```diff
@@ -345,18 +343,6 @@ func LaunchAsEndpoint(baseCtx context.Context,
 
-	if option.Config.IPAM == ipamOption.IPAMENI {
-		// ENI mode does not support IPv6.
-		if err := routingConfig.Configure(
```
Also, let's further remove any more references to ENI mode inside `routingConfig.Configure()` and make it clear it's only being used for Azure mode now.

twpayne replied:
We've now changed approach a bit here. We now use `routingConfig.Configure()` to set up the initial routes, and then `updateENIRulesAndRoutes` to handle changes.

christarazi commented on pkg/node/types/node.go:
```diff
@@ -151,6 +204,16 @@ func (n *RegisterNode) DeepKeyCopy() store.LocalKey {
+//
+// +k8s:deepcopy-gen=true
+type Interface struct {
+	Gateway           Address
```
Why does this need to be of an `Address` type? [1]

twpayne replied:
This is no longer needed.

christarazi commented on pkg/node/types/node.go:
```diff
@@ -448,3 +517,24 @@ func (n *Node) Unmarshal(data []byte) error {
+	subnetIP[len(subnetIP)-1]++
+
+	return Address{
+		Type: addressing.NodeInternalIP, // FIXME is this correct?
```
[1] would remove this FIXME

twpayne replied:
This is no longer needed.

christarazi reviewed (changes requested):
Given that this is a decent departure from how we've handled ENI before, it's worth going into the docs and updating it to reflect this new per-node model, instead of per-pod. 

I am also not clear on how new pods get their routing set up. How are we allocating new pod IPs?

twpayne committed:
logging: Add more standard fields

Signed-off-by: Tom Payne <tom@isovalent.com>

twpayne committed:
cidr: Add more test cases

Signed-off-by: Tom Payne <tom@isovalent.com>

gandro committed:
ipam: Pass MTU configuration to nodeStore

The IPAM CRD allocator will need to maintain the routes and rules on the
existing interfaces in cases new routes are added or removed. To be able
to do this, it will need to know the configured MTU.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro commented on pkg/datapath/linux/routing/eni.go:
```diff
@@ -0,0 +1,123 @@
+
+// ENIRulesAndRoutes returns the rules and routes required to configure an ENI
+// interface.
+func ENIRulesAndRoutes(
```
So this logic is not ENI specific anymore, right? It's basically just extracting the logic from `routingInfo.Configure` into a side-effect free function. We should probably call this something like `ComputeRulesAndRoutes` or something.

gandro commented on pkg/datapath/linux/routing/routing.go:
```diff
@@ -294,9 +248,9 @@ func retrieveIfIndexFromMAC(mac mac.MAC, mtu int) (index int, err error) {
+// ComputeTableIDFromIfaceNumber returns a computed per-ENI route table ID for the given
 // ENI interface number.
-func computeTableIDFromIfaceNumber(num int) int {
+func ComputeTableIDFromIfaceNumber(num int) int {
```
This can now be private again

gandro committed:
datapath/linux: Extract logic to compute rules and routes

This commit extracts the generation of rules and routes for ENI/Azure
IPAM interfaces from RoutingInfo.Configure and moves it into a separate
function.

This will allow subsequent commits to compute and compare the required
rules and routes in case new routing CIDRs are added or removed.

This commit contains no functional changes.

Co-authored-by: Tom Payne <tom@isovalent.com>
Signed-off-by: Tom Payne <tom@isovalent.com>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
aws/ec2: Exclude primary ENI IP from IPAM pool

This commit removes the primary ENI IP address from the IPAM pool. The
existing allocation code did already not consider it when computing the
number of IPs to allocate, but it was still returned in the allocation
result and therefore assigned to Cilium endpoints.

A subsequent commit will assign the primary IP to the Linux network
interface, to ensure SNAT works correctly for packets sent out via the
ENI. Therefore, we must not allocate the primary IP to endpoints.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
ipam: Dynamically update rules and routes for ENI

This commit extends the IPAM ENI control loop on the Cilium Agent to
detect for changes to the ENI VPC CIDRs and update the installed rules
and routes on each ENI accordingly. This allows users to maintain
connectivity between Cilium endpoints and remote targets which may
belong to a newly added secondary VPC CIDR.

In addition, this code also checks the CiliumNode CRD for newly added
interfaces and will assign the primary IP of the ENI to the
corresponding Linux network interface. This fixes an issue where SNAT
assigned the wrong IP to packets

Co-authored-by: Tom Payne <tom@isovalent.com>
Signed-off-by: Tom Payne <tom@isovalent.com>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
WIP: Disable dynamic route part

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
